subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c

```diff
@@ -235,58 +309,218 @@ struct pdu_adv *lll_adv_pdu_alloc(struct lll_adv_pdu *pdu, uint8_t *idx)
 
 	*idx = last;
 
-	p = (void *)pdu->pdu[last];
-	if (p) {
-		return p;
+	return adv_pdu_allocate(pdu, last);
+}
+
+struct pdu_adv *lll_adv_pdu_latest_get(struct lll_adv_pdu *pdu,
+				       uint8_t *is_modified)
+{
+	uint8_t first;
+
+	first = pdu->first;
+	if (first != pdu->last) {
+		uint8_t free_idx;
+		uint8_t pdu_idx;
+		void *p;
+
+		if (!MFIFO_ENQUEUE_IDX_GET(pdu_free, &free_idx)) {
+			LL_ASSERT(false);
+
+			return NULL;
```

Comment:
Well, this is a question to @cvinayak. I didn't add this function. Please take a look on master: https://github.com/zephyrproject-rtos/zephyr/blob/4d4a636555cdec8785652e016d974667be7c201b/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c#L272

Reply:
Its in the upstream code. I agree with @Thalley. @ppryga if its not too much, could you remove this `LL_ASSERT` and move to callers of `lll_adv_pdu_latest_get`

Reply:
@cvinayak sure I'll remove it

Reply:
Guys, I suggest to go with remove of return NULL. 
Please take a look e.g. here: https://github.com/zephyrproject-rtos/zephyr/blob/4d4a636555cdec8785652e016d974667be7c201b/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c#L910-L912

chan_prepare function hasn't got any way to handle it. If I add LL_ASSERT(!pdu); here then a code size increases. It will increase in every place I'll insert LL_ASSER. (I hope lto can do something about it). Besides that there is maintenance cost.

If I leave LL_ASSERT in `lll_adv_pdu_latest_get` then code size will not rise and there will be less maintenance.

Reply:
I've removed `return NULL` following reasoning from my last comment.

Reply:
Correct me if I'm wrong, but I don't think we should be concerned about code sizes from using LL_ASSERT: My understanding is that these would be compiled out for production builds.

Reply:
Yes, that is true. My bad, sorry. But... in such case, does it make sense to use LL_ASSERT in caller of the function? 
At the end of the day it will be removed. 

In my humble opinion the issue is still in the `lll_adv_pdu_latest_get` function (and other like this). 
If something very wrong had happen, then there wouldn't be a space to release past PDU (it would mean that we had a dangling pointer or more PDU than `pdu_free`  is able to store).
In such situation I would put here unconditional exception... 
If we leave it and return NULL then caller have to throw exception because in most cases there is no fall-back path.

Reply:
There are other functions that have similar issue e.g. `lll_adv_pdu_alloc`: https://github.com/zephyrproject-rtos/zephyr/blob/3cf64b17c35ea12c6d3fa49086d09374acc9149b/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c#L261-L269

Here we also have LL_ASSERT(p), but if `p` is `NULL` (there is no more free memory buffers for PDU) we are in a forest and cannot recover from that. 

What more it is not verified if the function returned NULL due to believe that it may not happen e.g. https://github.com/zephyrproject-rtos/zephyr/blob/3cf64b17c35ea12c6d3fa49086d09374acc9149b/subsys/bluetooth/controller/ll_sw/ull_adv.c#L1503-L1514

@Thalley you have opened pandora box ;-) 

@cvinayak What do you suggest to do with it?

Reply:
One reason to have the LL_ASSERT in the caller, rather than the callee, is that for debugging it makes it easier to figure out the context of the issue. 

It's very valid to find examples of where it's done like you added: Whatever the resolution is here, then perhaps that should be applied to other places afterwards.

Reply:
I've added changes as requested by @Thalley and @cvinayak. All changes are done as separate commit in PR.

I didn't add verification of return value for `lll_adv_pdu_alloc`. That would introduce more changes in callers and callers of callers... dure to use of `lll_adv_pdu_alloc` in functions that return `void`.